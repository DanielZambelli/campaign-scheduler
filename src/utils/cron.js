const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
