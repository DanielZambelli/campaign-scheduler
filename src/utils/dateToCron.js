const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
