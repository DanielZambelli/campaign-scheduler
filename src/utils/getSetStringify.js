const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
