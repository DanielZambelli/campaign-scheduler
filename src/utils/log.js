const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
