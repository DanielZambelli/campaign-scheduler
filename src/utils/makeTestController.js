const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
