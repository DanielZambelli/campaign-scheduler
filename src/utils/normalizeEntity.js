const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
