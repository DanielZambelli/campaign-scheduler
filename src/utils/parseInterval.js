const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
