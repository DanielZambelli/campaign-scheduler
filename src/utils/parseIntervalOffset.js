const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
