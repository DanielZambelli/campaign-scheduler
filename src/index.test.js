require('./utils/test')
const {seed} = require('./utils/test/seeds2')
const TEST = 'controller'

describe(TEST, () => {

  beforeEach(init(TEST,seed))

  afterEach(destroy)

  it('getActions', async () => {
    const res = await ctl.getActions().then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getActions', async () => {
    const res = await ctl.getActions({ subject: 'contact', subject: 'contact#4', actionId: 'action5' }).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getActions', async () => {
    const res = await ctl.getActions({ subject: 'contact', subject: 'contact#4', limit: 3 }).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getActions', async () => {
    const res = await ctl.getActions({ subject: 'contact', subject: 'contact#1', limit: 3 }).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getSchedules', async () => {
    const res = await ctl.getSchedules().then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getSchedules', async () => {
    const res = await ctl.getSchedules({limit: 3}).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getSchedules', async () => {
    const res = await ctl.getSchedules({limit: 2}).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getCampaigns', async () => {
    const res = await ctl.getCampaigns().then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getCampaigns', async () => {
    const res = await ctl.getCampaigns({ limit: 1 }).then(normalize)
    expect(res).toMatchSnapshot()
  })

  it('getCampaigns', async () => {
    const res = await ctl.getCampaigns({ active: false }).then(normalize)
    expect(res).toMatchSnapshot()
  })

})
