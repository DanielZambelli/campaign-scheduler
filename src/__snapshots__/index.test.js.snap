// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller getActions 1`] = `
{
  "hash": "9b2263ff5af20b3f0c3f9e920fa4b3d124829f73",
  "result": [
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate2",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action3",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate3",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate2",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action3",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate3",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate2",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "in a few seconds",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action3",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate3",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action4",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate4",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action5",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate5",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action6",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate6",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate2",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "in a few seconds",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action3",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate3",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action4",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate4",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action5",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate5",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action6",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate6",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getActions 2`] = `
{
  "hash": "6af8646e198824f5930891a39d4243d3161a697b",
  "result": [
    {
      "actionId": "action5",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate5",
      },
      "campaignId": "campaign3",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getActions 3`] = `
{
  "hash": "989db2448f309bfdd99b513f37c84b8f5794d2b5",
  "result": [],
}
`;

exports[`controller getActions 4`] = `
{
  "hash": "dbfc287d8a7ba2e132a2f5af17d20c3a46c53fc4",
  "result": [
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate2",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actionId": "action3",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate3",
      },
      "campaignId": "campaign1",
      "completedAt": null,
      "createdAt": "a few seconds ago",
      "expectedAt": "a few seconds ago",
      "log": null,
      "state": "pending",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getCampaigns 1`] = `
{
  "hash": "beb6212f6c9ad377794134c78b1ca554064e4168",
  "result": [
    {
      "actions": [
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate1",
          },
          "id": "action1",
          "interval": {
            "offset": "0seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate2",
          },
          "id": "action2",
          "interval": {
            "offset": "0seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate3",
          },
          "id": "action3",
          "interval": {
            "offset": "0seconds",
          },
        },
      ],
      "active": true,
      "createdAt": "a few seconds ago",
      "id": "campaign1",
      "updatedAt": "a few seconds ago",
    },
    {
      "actions": [
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate1",
          },
          "id": "action1",
          "interval": {
            "offset": "1hours",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate2",
          },
          "id": "action2",
          "interval": {
            "offset": "2hours",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate3",
          },
          "id": "action3",
          "interval": {
            "offset": "3hours",
          },
        },
      ],
      "active": false,
      "createdAt": "a few seconds ago",
      "id": "campaign2",
      "updatedAt": "a few seconds ago",
    },
    {
      "actions": [
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate1",
          },
          "id": "action1",
          "interval": {
            "offset": "0seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate2",
          },
          "id": "action2",
          "interval": {
            "offset": "1seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate3",
          },
          "id": "action3",
          "interval": {
            "offset": "1.1seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate4",
          },
          "id": "action4",
          "interval": {
            "offset": "1.2seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate5",
          },
          "id": "action5",
          "interval": {
            "offset": "1.3seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate6",
          },
          "id": "action6",
          "interval": {
            "offset": "1.4seconds",
          },
        },
      ],
      "active": true,
      "createdAt": "a few seconds ago",
      "id": "campaign3",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getCampaigns 2`] = `
{
  "hash": "80f9ff6893a8d47e8a1a3176eb1a0bf51371935d",
  "result": [
    {
      "actions": [
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate1",
          },
          "id": "action1",
          "interval": {
            "offset": "0seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate2",
          },
          "id": "action2",
          "interval": {
            "offset": "0seconds",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate3",
          },
          "id": "action3",
          "interval": {
            "offset": "0seconds",
          },
        },
      ],
      "active": true,
      "createdAt": "a few seconds ago",
      "id": "campaign1",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getCampaigns 3`] = `
{
  "hash": "16fde53aa036861a6f841cfdee36822e41e9853c",
  "result": [
    {
      "actions": [
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate1",
          },
          "id": "action1",
          "interval": {
            "offset": "1hours",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate2",
          },
          "id": "action2",
          "interval": {
            "offset": "2hours",
          },
        },
        {
          "callback": {
            "id": "sendEmail",
            "view": "emailTemplate3",
          },
          "id": "action3",
          "interval": {
            "offset": "3hours",
          },
        },
      ],
      "active": false,
      "createdAt": "a few seconds ago",
      "id": "campaign2",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getSchedules 1`] = `
{
  "hash": "9c0bd531b18b2f778c91104eb7bb979b66326909",
  "result": [
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": false,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#3",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": false,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#4",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign2",
      "createdAt": "a few seconds ago",
      "subject": "contact#5",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign2",
      "createdAt": "a few seconds ago",
      "subject": "contact#6",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign3",
      "createdAt": "a few seconds ago",
      "subject": "contact#7",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign3",
      "createdAt": "a few seconds ago",
      "subject": "contact#8",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getSchedules 2`] = `
{
  "hash": "d2bd3adb4323b87486227173a2891846e0dc3a5a",
  "result": [
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": false,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#3",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;

exports[`controller getSchedules 3`] = `
{
  "hash": "8a857ce6bff31dcdead8296a5a53c962ae07c63d",
  "result": [
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
    {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#2",
      "updatedAt": "a few seconds ago",
    },
  ],
}
`;
